File: String/leetcode_06.py

# Problem Link ====>> https://leetcode.com/problems/zigzag-conversion/description/

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s  # No change needed for a single row.

        rows = [[] for _ in range(min(numRows, len(s)))]
        currentRow = 0
        goingDown = False

        for char in s:
            rows[currentRow].append(char)

            if currentRow == 0 or currentRow == numRows - 1:
                goingDown = not goingDown

            currentRow += 1 if goingDown else -1

        result = ""
        for row in rows:
            result += "".join(row)

        return result


if __name__ == "__main__":
    s = "PAYPALISHIRING"
    numRows = 3

    solution = Solution()
    convertedString = solution.convert(s, numRows)

    print(f"The converted string is: {convertedString}")

File: String/leetcode_06.cpp

// Problem Link ====>> https://leetcode.com/problems/zigzag-conversion/description/

#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution
{
public:
    string convert(string s, int numRows)
    {
        if (numRows == 1)
        {
            return s; // No change needed for a single row.
        }

        vector<string> rows(min(numRows, int(s.length())));
        int currentRow = 0;
        bool goingDown = false;

        for (char c : s)
        {
            rows[currentRow] += c;

            if (currentRow == 0 || currentRow == numRows - 1)
            {
                goingDown = !goingDown;
            }

            currentRow += goingDown ? 1 : -1;
        }

        string result;
        for (const string &row : rows)
        {
            result += row;
        }

        return result;
    }
};

int main()
{
    string s = "PAYPALISHIRING";
    int numRows = 3;

    Solution solution;
    string convertedString = solution.convert(s, numRows);

    cout << "The converted string is: " << convertedString << endl;

    return 0;
}